Library/Homebrew/cask/lib/hbc/quarantine.rb

```diff
@@ -0,0 +1,70 @@
+require "development_tools"
+module Hbc
+  module Quarantine
+    module_function
+
+    QUARANTINE_ATTRIBUTE = "com.apple.quarantine".freeze
+
+    QUARANTINE_SCRIPT = "#{HOMEBREW_LIBRARY_PATH}/cask/lib/hbc/utils/quarantine.swift".freeze
+
+    # @private
+    def swift
+      @swift ||= DevelopmentTools.locate("swift")
+    end
+
+    def available?
+      status = !swift.nil?
+      odebug "Quarantine is #{status ? "available" : "not available"}."
+      !swift.nil?
+    end
+
+    def detect(file)
+      odebug "Verifying Gatekeeper status of #{file}"
+
+      quarantine_status = !status(file).empty?
+
+      odebug "#{file} is #{quarantine_status ? "quarantined" : "not quarantined"}"
+
+      quarantine_status
+    end
+
+    def status(file, command = SystemCommand)
+      command.run("/usr/bin/xattr",
+                  args:        ["-p", QUARANTINE_ATTRIBUTE, file],
+                  print_stderr: false).stdout
+    end
+
+    def cask(cask, downloaded_path, command = SystemCommand)
+      odebug "Quarantining #{downloaded_path}"
+      quarantiner = command.run(swift,
+                                args: [
+                                  QUARANTINE_SCRIPT,
+                                  downloaded_path,
+                                  cask.url.to_s,
+                                  cask.homepage.to_s,
+                                ])
+
+      return if quarantiner.success?
+
+      raise CaskError, "when quarantining #{downloaded_path}: #{quarantiner.stderr}"
+    end
+
+    def all(downloaded_path, base_path, command = SystemCommand)
+      odebug "Propagating quarantine status of #{downloaded_path}"
+
+      return unless detect(downloaded_path)
+
+      quarantine_status = status(downloaded_path)
+
+      base_path.each_child do |child|
+        next if detect(child)
+
+        quarantiner = command.run("/usr/bin/xattr",
+                                  args: ["-w", "-r", QUARANTINE_ATTRIBUTE, quarantine_status, child])
+
+        next if quarantiner.success?
+        raise CaskError, "when quarantining files in #{child}: #{quarantiner.stderr}"
```

Comment:
`"Failed to quarantine #{child}: #{quarantiner.stderr}"`

Reply:
Moved to its own `CaskQuarantineError` class.